bindings/rascal/utils/io.py

```diff
@@ -0,0 +1,376 @@
+        if file_extension == '.npy':
+            return True
+        else:
+            return False
+    else:
+        return False
+
+
+def get_class(module_name, class_name):
+    """Use module_name and class_name to make an instantiable class."""
+    module = importlib.import_module(module_name)
+    class_ = getattr(module, class_name)
+    return class_
+
+
+def obj2dict_beta(cls, state):
+    """Take a python object cls with its state and return a dictionary that
+    can be used to create a copy of this object.
+
+    Parameters
+    ----------
+    cls : object
+
+    state : dictionary
+        Contains the state of cls, i.e. the parameters used to initialize cls
+        in a 'init_params' field and the rest of the data needed to recover
+        the current state in a 'data' field.
+
+    Returns
+    -------
+    dictionary
+        fully serialized version of cls to a dictionary
+    """
+    VERSION = BETA_VERSION
+    module_name = cls.__module__
+    class_name = cls.__name__
+    frozen = dict(version=VERSION, class_name=class_name,
+                  module_name=module_name,
+                  init_params=state['init_params'],
+                  data=state['data'])
+    return frozen
+
+
+def dict2obj_beta(data):
+    """Take data, a dictionary created by the obj2dict function, and creates
+    a python object as described.
+
+    Parameters
+    ----------
+    data : dictionary
+        [description]
+
+    Returns
+    -------
+    deserialized python object described by data
+    """
+    cls = get_class(data['module_name'], data['class_name'])
+    obj = cls(**data['init_params'])
+    obj._set_data(data['data'])
+    return obj
+
+
+def is_valid_object_dict_beta(data):
+    """check compatibility of data to be used in dict2obj_beta"""
+    valid_keys = ['version', 'class_name', 'module_name', 'init_params', 'data']
+    aa = []
+    if isinstance(data, dict):
+        for k in data:
+            if k in valid_keys:
+                aa.append(True)
+        if len(aa) == len(valid_keys):
+            return True
+        else:
+            return False
+    else:
+        return False
+
+
+obj2dict = {BETA_VERSION: obj2dict_beta}
+dict2obj = {BETA_VERSION: dict2obj_beta}
+is_valid_object_dict = {BETA_VERSION: is_valid_object_dict_beta}
+
+
+def get_current_io_version():
+    return CURRENT_VERSION
+
+
+def get_supported_io_versions():
+    return list(dict2obj.keys())
+
+
+class BaseIO(ABC):
+    """
+    Definition of the interface of a python class that can be serialized by the
+    to_dict() function. It corresponds to 3 methods:
+
+    + _get_init_params is expected to return a dictionary containing all the
+    parameters used by the __init__() methods.
+
+    + _get_data is expected to return a dictionary containing all the data
+    that is not set by the initialization of the class.
+
+    + _set_data is expected to set the data that has been extracted by _get_data
+    """
+    @abstractmethod
```

Comment:
Nitpick here, but if the method implements default behaviour then it's not an abstract method (and this is not an abstract class). I recommend either removing the default behaviour (making everything `pass`), or better, removing the `@abstractmethod` from `_get_data()` and `_get_init_params()`, which would then let you remove the `_get_data()` overrides from the classes that don't need to implement it.

Reply:
the `@abstractmethod` wrapper is here to raise an error if the method is not implemented in the child class, so whatever happens in this methods is not used as a default and has to be called explicitly. So the fact that it is not `pass` is so that it is more convenient when one copy/paste these methods and writes the actual implementation an the actual class.

Reply:
That doesn't address the issue. If the method has default behaviour, then it should not have `@abstractmethod`. Why would one copy/paste if the method already has default behaviour defined in the base class? That's such an inefficient and error-prone way to do code re-use when you already have inheritance working for you here, not to mention it violates [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Reply:
you should tell to the people that implemented the abc module not me.

Reply:
Right, on going back and reading the PEP on ABCs, it seems it's mostly about defining standardized tests for whether a class _conforms_ to an interface. Strictly speaking, all this ABC stuff only serves to translate the call `isinstance(obj, BaseIO)`, e.g. here: https://github.com/cosmo-epfl/librascal/blob/e25ca3edd491052d85066a5dd82fec74c6ebbee1/bindings/rascal/utils/io.py#L237 into checking whether the object implements all the methods marked `@abstractmethod` in the `BaseIO` class.

So on reconsidering, the `@abstractmethod` marker here should definitely be kept, and my previous comments about this not being an abstract class were wrong (based on Java-style thinking). But to improve robustness, we should use the mechanism that the people who implemented the abc module designed to handle just this case: The `super()` mechanism (see the note at the bottom of [this doc page](https://docs.python.org/3/library/abc.html#abc.abstractmethod)). Every class that wants to use the default behaviour for e.g. the `_get_data()` method should instead write that method as

```python
def get_data():
    return  super()._get_data()
```

instead. It's more verbose but more robust (and definitely not "shiny"): That way, if we later decide to change the default behaviour (e.g. return some default keys, or throw an exception), it doesn't have to be changed in every implementing subclass.